src/convert.jl

```diff
@@ -1,14 +1,150 @@
-(::Type{SA})(x::Tuple{Tuple{Tuple{<:Tuple}}}) where {SA <: StaticArray} =
-    throw(DimensionMismatch("No precise constructor for $SA found. Length of input was $(length(x[1][1][1]))."))
+# A help wrapper to distinguish `SA(x...)` and `SA((x...，))`
+struct Args{T<:Tuple}
+    args::T
+end
+Length(x::Args) = Length(length(x.args))
+const BadArgs = Args{<:Tuple{Tuple{<:Tuple}}}
+
+# Some help functions.
+@pure has_ndims(::Type{<:StaticArray{<:Tuple,<:Any,N}}) where {N} = @isdefined N
+@pure has_ndims(::Type{<:StaticArray}) = false
+if VERSION < v"1.7"
+    Base.ndims(::Type{<:StaticArray{<:Tuple,<:Any,N}}) where {N} = N
+end
+@pure has_eltype(::Type{<:StaticArray{<:Tuple,T}}) where {T} = @isdefined T
+@pure has_eltype(::Type{<:StaticArray}) = false
+@pure has_size(::Type{<:StaticArray{S}}) where {S<:Tuple} = @isdefined S
+@pure has_size(::Type{<:StaticArray}) = false
+@pure has_size1(::Type{<:StaticMatrix{M}}) where {M} = @isdefined M
+@pure has_size1(::Type{<:StaticMatrix}) = false
+_size1(::Type{<:StaticMatrix{M}}) where {M} = M
+StaticSquareMatrix{N,T} = StaticMatrix{N,N,T}
+@generated function _sqrt(::Length{L}) where {L}
+    N = round(Int, sqrt(L))
+    N^2 == L || throw(DimensionMismatch("Input's length must be perfect square"))
+    return :($N)
+end
+
+const FirstClass = Union{SArray, MArray, SHermitianCompact, SizedArray}
+
+"""
+    construct_type(::Type{<:StaticArray}, x)
+
+Returns a constructor for a statically-sized array based on `x`'s size and eltype.
+The first argument is returned by default.
+"""
+function construct_type(::Type{SA}, x) where {SA<:StaticArray}
+    x isa BadArgs || return SA
+    _no_precise_size(SA, x.args[1][1])
+end
+
+# Here we define `construct_type(SA, x)` for `SArray`, `MArray`, `SHermitianCompact`, `SizedArray`
+# Different `x` has different rules, to summarize:
+# 1. Tuple
+#    We try to fill `SA` with elements in `x` if `SA` is static-sized.
+#    If `SA <: StaticVector`, the output `Length` is derived based on `length(x)`.
+#    If `SA <: StaticMatrix{M}`, the output `Size` is derived based on `length(x)÷M`.
+#    If `SA <: StaticMatrix{M,M} where M`, the output `Size` is derived based on `sqrt(length(x))`.
+#    If `length(SA) == 1 && length（x） > 1`, then we tries to fill `SA` with `x` itself. (rewrapping)
+# 2. Args (`SA(x...)`)
+#    Similar to `Tuple`, but rewrapping is not allowed.
+# 3. StaticArray
+#    Treat `x` as `Tuple` whenever possible. If failed, then try to inherit `x`'s `Size`.
+# 4. AbstractArray
+#    `x` is used to provide eltype. Thus `SA` must be static sized.
+function construct_type(::Type{SA}, x) where {SA<:FirstClass}
```

Comment:
I think this can be a docstring, instead of a comment. That change will be helpful in REPL. If we would like to keep this `construct_type` function internal, we can add warning in the docstring.
cf. https://github.com/SciML/ColPrac#changes-that-are-not-considered-breaking